### PR TITLE
REST API: Restrict Access to wc-blocks/v1 endpoint

### PR DESCRIPTION
**Pull request by timmyc**
This is a bit of a `try` branch to explore a possible way to restrict usage of `wc-blocks/v1` endpoints to "internal use" only. This approach, while not fully perfect, does restrict usage of the `wc-blocks/v1/products` endpoint to requests that have a local referring domain set.

I tested this out using the Featured Product block, but it should work for any block that utilizes the products endpoint. Here is a grab of the headers sent on the request from the editor:

image

So just adding a check to ensure the `referrer` includes the `admin_url` - if it does not, we throw an error. I'm likely using the incorrect HTTP code for this scenario, but just wanted to get this up for discussion more than anything.

__To Test__
- Open up the editor and add a featured product block
- Verify products are still displayed
- Perform an authenticated request using curl or Postman to the `wc-blocks/v1/products` endpoint and verify no results are returned.

__Thoughts__
This can obviously be circumvented by setting a referring header to include the admin_url - it is not "secure" in that manner, but it is more-or-less a simple deterrent to not try and use these endpoints outside the context of the editor. The existing permissions check is still in place to verify the user has the clearance to `edit_posts`.

### CODE REVIEW

File: includes/api/wc-blocks/class-wc-rest-blocks-products-controller.php [hunk 1]
```diff
@@ -81,6 +81,15 @@ public function get_items_permissions_check( $request ) {
 			return new WP_Error( 'woocommerce_rest_cannot_view', __( 'Sorry, you cannot list resources.', 'woocommerce' ), array( 'status' => rest_authorization_required_code() ) );
 		}
 
+		// Only allow requests to this endpoint to originate from local requests.
+		// This could obviously be spoofed, but will discourage usage outside of the editor.
+		$admin_url = admin_url();
+		$referrer = wp_get_referer();
+
+		if ( ! stristr( $referrer, $admin_url ) ) {
+			return new WP_Error( 'woocommerce_rest_cannot_view', __( 'This endpoint is for editor usage only. ' . $admin_url . '  ' . $referrer, 'woocommerce' ), array( 'status' => rest_authorization_required_code() ) );
```
Comment (ryelle) [inline includes/api/wc-blocks/class-wc-rest-blocks-products-controller.php]: I don't think you need the `$admin_url . '  ' . $referrer` here, the message alone should be fine

**Review by ryelle — COMMENTED** [review]
I talked to someone not-so-recently about how they have trouble with gutenberg because of CORS issues, since their backend and frontend are on two different domains– that wouldn’t be an issue here because you’re checking that admin referrer is on the admin domain, so that's fine…

What if we make a JS-powered block, and need to query the API from the frontend? I guess we'd need to change the user permissions too, so that's probably fine too.

This is working as described, if I set a fake `Referer` header it lets me see the data, without it I get the 403 error.

File: includes/api/wc-blocks/class-wc-rest-blocks-products-controller.php (hunk 1 above)
Comment (timmyc) [reply includes/api/wc-blocks/class-wc-rest-blocks-products-controller.php]: oh that was me debugging :blush:

**Comment by mikejolley** [comment]
@claudiosanches can you check this please?

**Comment by mikejolley** [comment]
If approved will need to target master as the blocks branch is merged.

**Comment by timmyc** [comment]
@mikejolley unless you are feeling pretty strongly that this approach is needed - I'm more in favor of closing this out, or punting to another release myself.

**Comment by mikejolley** [comment]
@timmyc I'm fine with docblock comments and a note in response headers/index that this is private.

**Comment by timmyc** [comment]
Closing in favor of #23007

**Pull request closed without merging**